src/eval.c

```diff
@@ -3605,6 +3617,304 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+	    ga_append(&plain, (int) **arg);
+	    continue;
+	}
+	else if (**arg == '$' && *(*arg + 1) == '$')
+	{
+	    // An escape of '$'
+	    ++*arg;
+	    ga_append(&plain, (int) **arg);
+	    continue;
+	}
+	else if (**arg == '$')
+	{
+	    char_u *stringified = (*(*arg + 1) == '{')
+		? enclose_template_expr_by_stringify(arg, is_literal_string)
+		: enclose_template_var_by_stringify(arg);
+	    if (stringified == NULL)
+	    {
+		ga_clear(&plain);
+		return FAIL;
+	    }
+
+	    ga_append(&plain, quote);            // '
+	    ga_concat(&plain, (char_u *) "..");  // ..
+	    ga_concat(&plain, stringified);      // stringify(x)
+	    ga_concat(&plain, (char_u *) "..");  // ..
+	    ga_append(&plain, quote);            // '
+
+	    vim_free(stringified);
+	    continue;
+	}
+	ga_append(&plain, (int) **arg);
+    }
+
+    if (**arg == NUL)
+    {
+	ga_clear(&plain);
+	semsg(_("E115: Missing quote: %s"), orig_arg);
+	return FAIL;
+    }
+    MB_PTR_ADV(*arg);
+    ga_append(&plain, quote);
+
+    if (!evaluate)
+    {
+	ga_clear(&plain);
+	return OK;
+    }
+    else
+    {
+	int	result;
+	char_u	*to_free = (char_u *) plain.ga_data;
+
+	result = eval1((char_u**) &plain.ga_data, rettv, TRUE);
+	vim_free(to_free);
+	return result;
+    }
+}
+
+    static int
+did_encount_string_quote(char_u **expr, int is_literal_string)
+{
+    // encounted a double quote in the single quoted string (lit str),
+    int did_encount_double_in_single =
+	(is_literal_string && **expr == '"');
+
+    // an escaped single quote in the single quoted string,
+    int did_encount_single_single_in_single =
+	(is_literal_string && **expr == '\'' && *(*expr + 1) == '\'');
+
+    // a single quote in the double quoted string,
+    int did_encount_single_in_double =
+	(!is_literal_string && **expr == '\'');
+
+    // or a single quote in the double quoted string?
+    int did_encount_escaped_double_in_double =
+	(!is_literal_string && **expr == '\\' && *(*expr + 1) == '"');
+
+    int result =
+	did_encount_double_in_single ||
+	did_encount_single_single_in_single ||
+	did_encount_single_in_double ||
+	did_encount_escaped_double_in_double;
+
+    if (did_encount_single_single_in_single || did_encount_escaped_double_in_double)
+    {
+	MB_PTR_ADV(*expr);
+    }
+
+    return result;
+}
+
+    static char_u*
+remove_all_quote_escaping(char_u *expr, int is_literal_string)
+{
+    garray_T	result;
+    char_u	*p;
+
+    ga_init2(&result, 1, 80);
+
+    for (p = expr; *p != NUL; MB_PTR_ADV(p))
+    {
+	int is_escaped_quote =
+	    (is_literal_string && *p == '\'' && *(p + 1) == '\'') ||
+	    (!is_literal_string && *p == '\\' && *(p + 1) == '"');
+	if (is_escaped_quote)
+	{
+	    ga_append(&result, *(p + 1));
+	    MB_PTR_ADV(p);
+	    continue;
+	}
+	ga_append(&result, *p);
+    }
+
+    return (char_u *) result.ga_data;
+}
+
+/*
+ * Parses a string like "${var}", and consume it onto "}".
+ *
+ * Return "{x->(type(x) is v:t_string?x:string(x)}(var)", or NULL if expr is invalid.
```

Comment:
s/(type(x)/type(x)/

I think it is better to insert spaces for readability.
`{x -> type(x) is v:t_string ? x : string(x)}(var)`

Reply:
Sorry, this is an oloder comment.
The latest implementation returns `"{x -> type(x) is v:t_string ? x : string(x)}(var)"`.
I fixed this :pray:

```c
 * Return "{x -> type(x) is v:t_string ? x : string(x)}(var)", or NULL if expr is invalid.
```